### PR TITLE
8319200: Don't use test thread factory in ProcessTools.createLimitedTestJavaProcessBuilder()

### PR DESCRIPTION
**Pull request by lmesnik**
Test thread factory is a mode similar to VM flags and should not be used in ProcessTools.createLimitedTestJavaProcessBuilder(). Only createTestJavaProcessBuilder() should use it like jtreg VM options.

Adding the test thread factory requires the injection of arguments in the middle of the list. I don't think it makes sense to modify arguments in several places so I replaced it with the flag isLimited and moved all modifications in createJavaProcessBuilder().

Testing tier1-5.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319200](https://bugs.openjdk.org/browse/JDK-8319200): Don't use test thread factory in ProcessTools.createLimitedTestJavaProcessBuilder() (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16442/head:pull/16442` \
`$ git checkout pull/16442`

Update a local copy of the PR: \
`$ git checkout pull/16442` \
`$ git pull https://git.openjdk.org/jdk.git pull/16442/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16442`

View PR using the GUI difftool: \
`$ git pr show -t 16442`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16442.diff">https://git.openjdk.org/jdk/pull/16442.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16442#issuecomment-1788205255)

### CODE REVIEW

File: test/lib/jdk/test/lib/process/ProcessTools.java [hunk 1]
```diff
@@ -444,10 +444,11 @@ private static List<String> addTestThreadFactoryArgs(String testThreadFactoryNam
     /**
      * Create ProcessBuilder using the java launcher from the jdk to be tested.
      *
+     * @param isLimited Defines if jtreg options and test thread factory should be added
```
Comment (dholmes-ora) [inline test/lib/jdk/test/lib/process/ProcessTools.java:447]: Suggestion:
```
@param isLimited If true, prevents the addition of jtreg options to the command, and ignores any available test thread factory
```

**Review by dholmes-ora — COMMENTED** [review]
> Test thread factory is a mode similar to VM flags and should not be used in ProcessTools.createLimitedTestJavaProcessBuilder(). 

I'm not sure I agree with that. I think the test factory is a separate testing dimension, independent of JVM flags.

**Review by stefank — CHANGES_REQUESTED** [review]
It seems like the code would be cleaner if you moved the threadFactory injection to `createTestJavaProcessBuilder`.

File: test/lib/jdk/test/lib/process/ProcessTools.java [hunk 2]
```diff
@@ -444,10 +444,11 @@ private static List<String> addTestThreadFactoryArgs(String testThreadFactoryNam
...
      * @param command Arguments to pass to the java command.
      * @return The ProcessBuilder instance representing the java command.
      */
-    private static ProcessBuilder createJavaProcessBuilder(String... command) {
+    private static ProcessBuilder createJavaProcessBuilder(boolean isLimited, String... command) {
```
Comment (msheppar) [inline test/lib/jdk/test/lib/process/ProcessTools.java:451]: addThreadFactory might be a more appropriate name
However, Stefan has a more SOLID suggestion

**Comment by lmesnik** [comment]
Moved the threadFactory injection to createTestJavaProcessBuilder. I think that it is more logical to don't set it for limited version of process builder.

File: test/lib/jdk/test/lib/process/ProcessTools.java [hunk 3]
```diff
@@ -384,7 +384,7 @@ public static long getProcessId() throws Exception {
       java <jvm-args> -Dtest.thread.factory=<test-thread-factory-name> jdk.test.lib.process.ProcessTools <test-thread-factory-name> <test-class> <test-args>
      */
 
-    private static List<String> addTestThreadFactoryArgs(String testThreadFactoryName, List<String> command) {
+    private static String[] addTestThreadFactoryArgs(String testThreadFactoryName, String[] command) {
```
[the remark this answers is not in the document]
Comment (lmesnik) [reply test/lib/jdk/test/lib/process/ProcessTools.java]: done

File: test/lib/jdk/test/lib/process/ProcessTools.java [hunk 4]
```diff
@@ -398,7 +398,7 @@ private static List<String> addTestThreadFactoryArgs(String testThreadFactoryNam
 
         boolean expectSecondArg = false;
         boolean isTestThreadFactoryAdded = false;
-        for (String cmd : command) {
+        for (String cmd : commandArgs) {
```
Comment (dholmes-ora) [inline test/lib/jdk/test/lib/process/ProcessTools.java:401]: So `cmd` should really be `arg` right?

**Comment by dholmes-ora** [comment]
I remain concerned that this means that a whole swag of tests will never be run with virtual threads, which reduces our virtual thread test coverage. Hard to quantify. Do you have any stats on how many tests this will affect and which ones?

**Comment by msheppar** [comment]
> I remain concerned that this means that a whole swag of tests will never be run with virtual threads, which reduces our virtual thread test coverage. Hard to quantify. Do you have any stats on how many tests this will affect and which ones?

I thought the purpose of the createLimitedTestJavaProcessBuilder is for use where vm option are NOT  being propagated to the ProcessBuilder,  as such the thread factory property wont come into consideration ? the Java doc notes that test using createLimitedTestJavaProcessBuilder would be marked as vm.flagless which would suugest thread factory is not being used.
createTestJavaProcessBuilder is for use when vm args are being propagated and as such thread factory can be "injected" via this method. So this change is streamlining these use cases.
Is this a correct interpretation?

**Comment by lmesnik** [comment]
The createLimitedTestJavaProcessBuilder is used in 410 tests (hotspot and jdk).
233 of them are flagless and not supposed to be executed with any additional VM flags. (They should be reviewed by it is a separate issue).
177 are not marked as flagless and should be updated to be flagless or use createTestJavaProcessBuilder.
The 'createLimitedTestJavaProcessBuilder' method should be used only when the process not accept any flags and it is logical to assume that thread factory shouldn't be used either.  I think it just makes consistent the createLimitedTestJavaProcessBuilder and createTestJavaProcessBuilder methods.
It was not my original goal to add test thread factory part of createLimitedTestJavaProcessBuilder, I just missed that original createJavaProcessBuilder is applied to all processes.

**Comment by dholmes-ora** [comment]
Okay, sorry, I've misunderstood the nature of this issue. This change has no affect on whether jtreg runs a specific test (or set of tests) using the virtual thread wrapper to launch the test. This only affects code that uses ProcessTools and where the launched JVM would examine the `test.thread.factory` property. How many tests actually look at this property directly in code that is run via ProcessTools? The only tests that appear to look at this property are JDB and JDI tests, where we have a test framework that selects use of virtual threads based on this property. I couldn't find any test that would actually be affected by this. Sorry for the noise.

**Comment by dholmes-ora** [comment]
Uggghhh no I've still misunderstood what this is doing. `addTestThreadFactoryArgs` will create a wrapper to run the specified main class in a virtual thread. So if the test using `createLimitedTestJavaProcessBuilder` was being run in a virtual thread, then the exec'd "app" would be too. Plus it would have the property set. So this is potentially reducing virtual thread test coverage.

That said I'm bowing out of this discussion.

**Comment by lmesnik** [comment]
It is expected from `createLimitedTestJavaProcessBuilder` to execute the process exactly as specified in the test, without any additional changes from jtreg. It is very likely that the test might fail if the process is executed in other modes. I agree that usage of this method is always a potential loss of coverage for vm flags being tested and for virtual threads. So this method should be used only when it is necessary. As well as we should minimize the number of flagless testing.

**Pull request closed without merging**